### PR TITLE
ScriptService: Replace max compilation per minute setting with max compilation rate

### PR DESCRIPTION
**Pull request by spinscale**
The current script service has a script compilation limit for a one
minute window. This is set to a small default value of 15. Instead of
increasing that default value, this commit enhances the time window to
five minutes, so that the script service can deal with bursts by being
able to compile the number of scripts allowed to compile in five minutes
also in one.

Also the setting has been renamed.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 1]
```diff
@@ -66,8 +66,8 @@
         Setting.positiveTimeSetting("script.cache.expire", TimeValue.timeValueMillis(0), Property.NodeScope);
     public static final Setting<Integer> SCRIPT_MAX_SIZE_IN_BYTES =
         Setting.intSetting("script.max_size_in_bytes", 65535, Property.NodeScope);
-    public static final Setting<Integer> SCRIPT_MAX_COMPILATIONS_PER_MINUTE =
-        Setting.intSetting("script.max_compilations_per_minute", 15, 0, Property.Dynamic, Property.NodeScope);
+    public static final Setting<Integer> SCRIPT_MAX_COMPILATIONS_RATE =
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: I'm not sure this new name is better. It has no indication what the base of the rate is? One must find docs to know what a value means, while before it was very clear from the name (per minute). Can we have something to indicate this, or have the values not be a plain integer, but instead a string with some kind of unit? I realize this is creeping the scope a bit, but I think this new setting will just be confusing (eg look at the history of size values without units in 1.x days).

Comment (spinscale) [reply core/src/main/java/org/elasticsearch/script/ScriptService.java]: I agree, that it should be obvious by reading the setting how it works.

I changed the setting to be a real rate, which consists of an nonnegative integer and a positive timevalue split by a `/` defaulting to `75/5m`. Is this what you were after?

File: core/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 2]
```diff
@@ -60,14 +61,37 @@
 
     static final String DISABLE_DYNAMIC_SCRIPTING_SETTING = "script.disable_dynamic";
 
+    // a parsing function that requires a non negative int and a timevalue as arguments split by a slash
+    // this allows you to easily define rates
+    static final Function<String, Tuple<Integer, TimeValue>> MAX_COMPILATION_RATE_FUNCTION =
+            (String value) -> {
+                if (value.contains("/") == false || value.startsWith("/") || value.endsWith("/")) {
+                    throw new IllegalArgumentException("parameter must contain a positive number and a timevalue, i.e. 10/1m");
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: This should contain the value that failed validation?

File: core/src/test/java/org/elasticsearch/script/ScriptServiceTests.java [hunk 3]
```diff
@@ -82,30 +87,47 @@ StoredScriptSource getScriptFromClusterState(String id) {
...
         }
     }
 
+    public void testMaxCompilationRateSetting() throws Exception {
+        assertThat(MAX_COMPILATION_RATE_FUNCTION.apply("10/1m"), is(Tuple.tuple(10, TimeValue.timeValueMinutes(1))));
+        expectThrows(ElasticsearchParseException.class, () -> MAX_COMPILATION_RATE_FUNCTION.apply("10/m"));
+        expectThrows(ElasticsearchParseException.class, () -> MAX_COMPILATION_RATE_FUNCTION.apply("6/1.6m"));
+        expectThrows(NumberFormatException.class, () -> MAX_COMPILATION_RATE_FUNCTION.apply("foo/bar"));
+        expectThrows(NumberFormatException.class, () -> MAX_COMPILATION_RATE_FUNCTION.apply("6.0/1m"));
+        expectThrows(IllegalArgumentException.class, () -> MAX_COMPILATION_RATE_FUNCTION.apply("6/-1m"));
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/script/ScriptServiceTests.java]: I think we should be checking all of these expected exceptions. We use IllegalArgumentException all over, so we should be double checking which one. Even within the rate parsing, we throw 3 different IAEs.

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java [hunk 4]
```diff
@@ -337,7 +337,7 @@ public InternalTestCluster(long clusterSeed, Path baseDir,
         builder.put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING.getKey(), "1b");
         builder.put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING.getKey(), "1b");
         // Some tests make use of scripting quite a bit, so increase the limit for integration tests
-        builder.put(ScriptService.SCRIPT_MAX_COMPILATIONS_PER_MINUTE.getKey(), 1000);
+        builder.put(ScriptService.SCRIPT_MAX_COMPILATIONS_RATE.getKey(), "1000/1m");
```
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: As a follow up, I wonder if we really need to support this enormous rate, or if the tweaks to the default might now be sufficient for test clusters?

File: core/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 5]
```diff
@@ -60,14 +61,37 @@
...
+                int idx = value.indexOf("/");
+                String count = value.substring(0, idx);
+                String time = value.substring(idx + 1);
+
+                int rate = Integer.parseInt(count);
+                if (rate < 0) {
+                    throw new IllegalArgumentException("rate ["+ rate +"] must be positive");
+                }
+                TimeValue timeValue = TimeValue.parseTimeValue(time, "script.max_compilations_rate");
+                if (timeValue.nanos() <= 0) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/script/ScriptService.java]: I wonder if we should have a minimum time value. For example, if someone does a rate per 1s, this is something we are unlikely able to handle reliably.  Perhaps the minimum should be per 1 minute?

**Review by rjernst — COMMENTED** [review]
I left a couple more minor notes.

File: core/src/main/java/org/elasticsearch/script/ScriptService.java (hunk 1 above)
Comment (rjernst) [reply core/src/main/java/org/elasticsearch/script/ScriptService.java]: Great, exactly what I was hoping for.

File: core/src/test/java/org/elasticsearch/script/ScriptServiceTests.java (hunk 3 above)
Comment (spinscale) [reply core/src/test/java/org/elasticsearch/script/ScriptServiceTests.java]: done

File: core/src/main/java/org/elasticsearch/script/ScriptService.java (hunk 5 above)
Comment (spinscale) [reply core/src/main/java/org/elasticsearch/script/ScriptService.java]: ++, set it to 1 minute

File: core/src/main/java/org/elasticsearch/script/ScriptService.java (hunk 2 above)
Comment (spinscale) [reply core/src/main/java/org/elasticsearch/script/ScriptService.java]: indeed. done.

File: test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java (hunk 4 above)
Comment (spinscale) [reply test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java]: yeah, trying to stick with the new default by default might make most sense... I'll give it a go tomorrow along with some test runs.

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**